### PR TITLE
add:create new branch with updated files

### PR DESCRIPTION
**Pull request by mertakman**
This PR introduces an automated workflow to streamline the process of updating Microsoft Go versions for Azure Linux. Whenever a new Microsoft Go release is published, the following actions are automatically triggered:
Updating the specfile with changelog,
Updating the download URL and version in cgmanifest.json,
Updating the signature hash and download URL link in signatures.json.

### CODE REVIEW

File: cmd/releasego/update-azure-linux.go
Comment (dagood) [inline cmd/releasego/update-azure-linux.go:138]: This would make a separate commit per file, right? Could we use https://pkg.go.dev/github.com/google/go-github/github#GitService.CreateTree -> CreateCommit -> UpdateRef instead? (I've done it before, but from C# code.)

File: cmd/releasego/update-azure-linux.go
Comment (dagood) [inline cmd/releasego/update-azure-linux.go:120]: Will this fail if the ref already exists? Maybe we should create a new ref instead? (In case of a rerun.)

File: cmd/releasego/update-azure-linux.go
Comment (dagood) [inline cmd/releasego/update-azure-linux.go:153]: FWIW: I think the Azure Linux team will be fine with us having custom info here, not necessarily something based on the template.

File: cmd/releasego/update-azure-linux.go
Comment (dagood) [inline cmd/releasego/update-azure-linux.go:112]: Nit: errors should start with lowercase.

File: cmd/releasego/update-azure-linux.go
Comment (dagood) [inline cmd/releasego/update-azure-linux.go:139]: Reuse PR title func?

File: cmd/releasego/update-azure-linux.go
Comment (dagood) [reply cmd/releasego/update-azure-linux.go:120]: Talked about this in sync: letting it fail helps avoid potentially generating more PRs/branches than we want, and we will have a parameter on the release pipeline to let us specify a new name upon rerun if we truly want a new branch/PR. This also preserves history rather than overwrite an existing branch/PR.

File: cmd/releasego/update-azure-linux.go
Comment (dagood) [inline cmd/releasego/update-azure-linux.go:381]: The idea of this general area is to preserve the order of the original keys, right? (Which is something I think we do need to do--to avoid thrashing with other bots (or opinionated humans).)

I think since `registration` is a map, this line ends up sorting it again anyway.

Maybe we could implement something like https://github.com/golang/go/issues/27179#issuecomment-587528269 to preserve order without quite as much complication?

It could also be simpler to treat the file as only text rather than actually fully parsing it as JSON, so it's easier to leave most of the file untouched no matter what other people do to the file.

Comment (mertakman) [reply cmd/releasego/update-azure-linux.go:395]: > The idea of this general area is to preserve the order of the original keys, right? (Which is something I think we do need to do--to avoid thrashing with other bots (or opinionated humans).)
> 
> I think since `registration` is a map, this line ends up sorting it again anyway.
> 
> Maybe we could implement something like [golang/go#27179 (comment)](https://github.com/golang/go/issues/27179#issuecomment-587528269) to preserve order without quite as much complication?
> 
> It could also be simpler to treat the file as only text rather than actually fully parsing it as JSON, so it's easier to leave most of the file untouched no matter what other people do to the file.

So I did the another thing mentioned in same issue:
https://github.com/golang/go/issues/27179#issuecomment-848662008

Comment (dagood) [reply cmd/releasego/update-azure-linux.go:411]: My worry here is that a bot will change the key order back to what it was, no matter how minimal the difference is. Then next time we run the update, our bot will change it back to "our" order, and then they'll swap it back, and so on. (And, not necessarily an Azure Linux bot: any other team doing auto-updates too.)

Ultimately this just means things are harder to review/`blame` and the Git repo might get a tiny bit larger, so I suppose it's up to Azure Linux. And, if they really want to avoid this problem, they need to add CI that verifies that this file is laid out in a prescriptive order.

File: cmd/releasego/update-azure-linux.go
Comment (gdams) [inline cmd/releasego/update-azure-linux.go:192]: This needs to be a variable

Comment (mertakman) [reply cmd/releasego/update-azure-linux.go]: fixed, thanks for mentioning.

**Comment by gdams** [comment]
@mertakman the tests are failing now

**Comment by mertakman** [comment]
> @mertakman the tests are failing now

Fixed

File: cmd/releasego/update-azure-linux.go
Comment (gdams) [inline cmd/releasego/update-azure-linux.go:436]: remove this

File: cmd/releasego/update-azure-linux.go
Comment (mertakman) [reply cmd/releasego/update-azure-linux.go:147]: After discussion with azure linux team member we decided to go with squash and merge, this removes a  need for making everything in single commit.

File: cmd/releasego/update-azure-linux.go
Comment (mertakman) [reply cmd/releasego/update-azure-linux.go]: Fixed with using strudt instead of map for json marshalling.

**Review by gdams — APPROVED** [review]
LGTM although I'll let @dagood give it another review before merging

File: cmd/releasego/update-azure-linux.go
Comment (dagood) [inline cmd/releasego/update-azure-linux.go:193]: `Sprintf` calls this `format`, which I'd prefer--calling it `template` implies text/template to me.

Comment (dagood) [inline cmd/releasego/update-azure-linux.go:193]: `const template =` rather than `:=`, to allow arg mismatches to be detected by `go vet` and the linter.

File: cmd/releasego/update-azure-linux.go
Comment (dagood) [inline cmd/releasego/update-azure-linux.go:154]:
```suggestion
			return fmt.Errorf("failed to update file %q: %v", filePath, err)
```

A few more capital errors earlier in the file, too.

File: cmd/releasego/update-azure-linux.go
Comment (dagood) [inline cmd/releasego/update-azure-linux.go:398]: Why change away from the pointer?

File: cmd/releasego/update-azure-linux.go
Comment (gdams) [reply cmd/releasego/update-azure-linux.go:52]: Thinking aloud here, does the buddy build need the branch or the PR to exist in order to run? We might be able to create the branch and push the code, run the buddy build and then raise the PR afterwards?

File: cmd/releasego/update-azure-linux.go
Comment (mertakman) [reply cmd/releasego/update-azure-linux.go:196]: I started a Teams thread about this to discuss further. Let's see where it evolves and then we can apply the change here.

File: cmd/releasego/update-azure-linux.go
Comment (dagood) [reply cmd/releasego/update-azure-linux.go]: The PR specifically, currently. In the past, we've been told directly that we need to submit the PR first and then edit in the build ID.

Here's the parameter doc for the pipeline:

https://dev.azure.com/mariner-org/mariner/_build?definitionId=2190

> image

Our branch is in a fork, so (unless there's some secret mechanism--which may be unintentional as far as Azure Linux is concerned) we can't specify our branch directly as a topic branch.

**Pull request merged**
